packages/migration/src/migrations/v0042_removeTrackName.ts

```diff
@@ -14,7 +14,38 @@
  * limitations under the License.
  */
 
-function removeTrackName(storyData) {
+/**
+ * Internal dependencies
+ */
+import type { ElementV43, PageV43, StoryV43 } from './v0043_removeTagNames';
```

Comment:
@barklund @swissspidy Before I continue with all the files, wanted to ask for feedback for this approach (see this file and the next, `v0043_` file).

The idea is that the last migration returns the current `Story` type, however, the migration before that returns `StoryV43` type, and the one before that `StoryV42` type, etc. Each migration will then extend the returned story type, e.g. the last migration returns `Story` as it currently is but it applies the migration script on an extended `Story` type -- as it was before migration.

Typed these two files as an example for now. Thoughts?

Reply:
I like it. It's super clean and type-safe. I've checked around, but don't see any recommendations as to what to do in general, so this is as good as any approach IMO.

But let's see what @swissspidy says too.

Reply:
Hmm the downside with that is whenever you add a new migration, you have to edit the previous one as well.

Looking at this file here (`v0042_removeTrackName`), intuitively I would think that `removeTrackName` takes a `StoryV41` type (from the previous migration) and returns a new `StoryV42` type (the resulting version after this migration).

The `migrate()` function then is the one returning the current `Story` type.

WDYT?

Note: Not sure about `v0004`, which has two migrations (`squareToShape` and `mediaElementToResource`)

Reply:
> Hmm the downside with that is whenever you add a new migration, you have to edit the previous one as well.

Yep, unfortunately. Not sure if there's a better option though.

> Looking at this file here (v0042_removeTrackName), intuitively I would think that removeTrackName takes a StoryV41 type (from the previous migration) and returns a new StoryV42 type (the resulting version after this migration).
The migrate() function then is the one returning the current Story type.

Sure, sounds good. The last migration would still also return the current Story type. Although we could still create a new type for clarity / to have to make less changes for the next migration.

> Note: Not sure about v0004, which has two migrations (squareToShape and mediaElementToResource)

Could have StoryV4_0 and StoryV4_1 based on the index in the array of migrations for example.

Reply:
@swissspidy I tried so that StoryV40 is defined in the `v40` file and then `v41` migration imports it from `v40` to take that type as the input, however, since the changes that happen in V41 are then type in V40 file this way, it gets quite confusing to read this -- the changes happening in each file have nothing to do with the type that's defined in that file. It might be correct by the numbers but it's quite confusing to read.

I currently changed the code so that now the input version is defined in the file that actually uses it, PTAL. E.g. V39 is defined in V40 file and used there as well and relates to the code there. It's much easier to read this way but now importing V39 is from V40 file, which is also a bit confusing. It's still easier to read though, so 🤷 

Let me know if you still think it's better to have it correct in terms of numbers (defining `StoryV39` in `V39` file but the relevant code actually being in V40).

Reply:
> Although we could still create a new type for clarity / to have to make less changes for the next migration.

👍 💯  

-----

> I tried so that StoryV40 is defined in the `v40` file and then `v41` migration imports it from `v40` to take that type as the input, however, since the changes that happen in V41 are then type in V40 file this way, it gets quite confusing to read this

I am definitely confused now. Did you push these changes? Because that's not what I am seeing.

And why are V41 types in V40? V40 should not know anything about future migrations, because they don't exist yet at the time of writing.

> I currently changed the code so that now the input version is defined in the file that actually uses it, PTAL. E.g. V39 is defined in V40 file and used there as well and relates to the code there.

Why is V39 defined in V40 and not in the actual V39 file? That seems like the same setup as yesterday, where one has to modify old migrations whenever you need a new one.

----

What I had in mind when I wrote my comment yesterday:

A V39 migration:

1. Imports V38 type from V38 file
2. Defines a new V39 type
3. Takes V38 data and returns V39 data
4. Exports V39 type so it can be used by the next migration

This way, V39 builds upon the previous migration and has no awareness of any future migrations like V40

Reply:
> This way, V39 builds upon the previous migration and has no awareness of any future migrations like V40

Yep, I understood what you meant -- I tried it out and it was just quite hard to read + one still has to modify old migrations whenever we need one because we don't know what will change in the future. The last migration's types are the same as the current type but if anything changes in the current type, we'll still need to change the previous migration's types as well.

Anyway -- I can definitely still do as you recommend, will update on Monday.

Reply:
> This way, V39 builds upon the previous migration and has no awareness of any future migrations like V40

Actually there's a problem with this as well -- we don't know how the original Story type looked like in the first migration and we don't know which properties were added while each next migration version was added. Migration doesn't migrate ALL the changes, just the ones that need migration.

> And why are V41 types in V40? V40 should not know anything about future migrations, because they don't exist yet at the time of writing.

Because we need a base story type to take as the reference type. Since we don't have the very original Story of version 0 as the reference and we don't know which properties were added meanwhile to fully write out every type, we can't take the version 0 Story as the reference and start extending from Version 0. We can instead do the other way around -- the last migration takes the current Story type as the reference every time and start extending from the last migration. When a new migration is added, the previously last one takes the new migration's story type as the reference.

If V39 builds upon the previous migration then we would need the Story type of V0 to start from there.

Let me know if I explained this clearly, perhaps you have an easy way in mind to get the Story type for V0 to take as the basis.

Reply:
Sounds like I might need to look into the code myself to better understand the drawbacks. Definitely want to avoid unnecessary back and forth here, so let's first make sure we are all on the same page.

From the looks of it, it definitely sounds like we need to work backwards and create a V0 type that is then used by the first migration.

Reply:
> From the looks of it, it definitely sounds like we need to work backwards and create a V0 type that is then used by the first migration.

We could do it but probably just based on the changes in the migration and not the "real" story type that actually existed in the beginning since when adding new props we don't do migration.

However, when taking the last Story as the base, we would end up with the same result since every version type is considering just the changes reflected in the migration files. 

If that's acceptable then we could definitely do that, just take the current Story type and work back through the migrations to create the initial Story type (which would not be fully accurate but reflect the migrations correctly). Thoughts?